File: java/0141_hasCycle.java
/**
 * Definition for singly-linked list.
 * class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode(int x) {
 *         val = x;
 *         next = null;
 *     }
 * }
 */
public class Solution {
    public boolean hasCycle(ListNode head) {
        ListNode p1=head,p2=head;
        while(p2!=null && p2.next!=null)
        {
            if(p2.next==p1)
                return true;
            p2 = p2.next.next;
            p1 = p1.next;
        }
        return false;    
    }
}

File: cpp/0141_hasCycle.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
// 环链表，快慢指针
// 快指针如果遇到null，则返回false
// 如果快指针遇到慢指针，返回true
class Solution {
public:
    bool hasCycle(ListNode *head) {
        auto p1=head,p2=head;
        if(p1==NULL)
            return false;
        else
            p2=head->next;
        while(p2!=NULL && p2->next!=NULL)
        {
            if(p2==p1)
                return true;
            p2 = p2->next->next;
            p1 = p1->next;
        }
        return false;
    }
};
